presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java

```diff
@@ -134,7 +138,15 @@ MemoryTrackingContext getQueryMemoryContext()
 
     private synchronized ListenableFuture<?> updateSystemMemory(long delta)
     {
+        // We get the user memory from the pool instead of calling queryMemoryContext.getUserMemory()
+        // as the latter may cause a deadlock.
+        // A thread doing user allocation will lock the monitors of RootAggregatedMemoryContext (queryMemoryContext.userAggregateMemoryContext) -> QueryContext
+        // and a thread calling updateSystemMemory will lock QueryContext -> RootAggregatedMemoryContext (queryMemoryContext.userAggregateMemoryContext)
```

Comment:
I would write it this way:

```
// We call memoryPool.getQueryUserMemoryReservation(queryId) instead of queryMemoryContext.getUserMemory().
//
// Calling the latter would result in dead lock:
// * A thread doing user allocation would acquire locks in this order:
//   1. monitor of queryMemoryContext.userAggregateMemoryContext
//   2. monitor of this (QueryContext)
// * this thread would acquire locks in this order:
//   1. monitor of this (QueryContext)
//   2. monitor of queryMemoryContext.userAggregateMemoryContext